aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -0,0 +1,274 @@
+    thrust::copy_n(policy, srcs, n, dsts);
+  });
+}
+
+// cufft
+Tensor _fft_cufft(const Tensor& self, int64_t signal_ndim,
+                  bool complex_input, bool complex_output,
+                  bool inverse, IntList checked_signal_sizes,
+                  bool normalized, bool onesided,
+                  IntList output_sizes) {
+  Tensor input = self;
+
+  bool is_half = input.type().scalarType() == ScalarType::Half;
+
+  // cuFFT requires input and output data pointers to complex type aligned
+  // our allocated output tensor is always 256 bytes aligned so it is fine, but
+  // we need to check input tensor to make sure that it is not unaligned, e.g.
+  // from a slicing.
+  auto complex_size_bytes = 2 * input.type().elementSizeInBytes();
+  if (reinterpret_cast<std::uintptr_t>(input.data_ptr()) % complex_size_bytes != 0) {
+    input = self.clone();
+  }
+
+  // check input batch size
+  long long int batch = input.size(0);
+
+  // check the input sizes and strides to see if we need to make it contiguous
+  std::vector<long long int> signal_sizes(signal_ndim);
+  std::vector<long long int> inembeds(signal_ndim);
+  long long int istride, signal_size;
+  bool need_contiguous = input.stride(signal_ndim) == 0;
+  if (complex_input) {
+    // real/imag dimension must be like complex type
+    need_contiguous |= input.stride(-1) != 1;
+  } else if (is_half) {
+    // For half, base strides on the real part of real-to-complex and
+    // complex-to-real transforms are not supported. Since our output is always
+    // contiguous, only need to check real-to-complex case.
+    need_contiguous |= input.stride(signal_ndim) != 1;
+  }
+  // store last tensor stride to infer inembeds array
+  // complex input's last dim is size 2 and contiguous
+  long long int ilast_stride = complex_input ? 2 : 1;
+   // for each signal dim from innermost to outermost
+  for (int64_t i = signal_ndim - 1; i >= 0; i--) {
+    signal_size = checked_signal_sizes[i];
+    istride = input.stride(i + 1);
+    if (is_half && !is_pow_of_two(signal_size)) {
+      std::ostringstream ss;
+      ss << "cuFFT doesn't support signals of half type with size that is not "
+         << "a power of two, but gets a signal size of " << signal_size << " at"
+         << "signal dimension " << i;
+      throw std::runtime_error(ss.str());
+    }
+    signal_sizes[i] = signal_size;
+    if (!need_contiguous) {
+      // set the inembeds for dim in last dimension
+      if (ilast_stride == 0) {
+        need_contiguous = istride != 0;
+        if (i < signal_ndim - 1) {
+          inembeds[i + 1] = 0;
+        }
+      } else {
+        need_contiguous = istride % ilast_stride != 0;
+        if (i < signal_ndim - 1) {
+          inembeds[i + 1] = istride / ilast_stride;
+        }
+      }
+      ilast_stride = istride;
+    }
+  }
+
+  need_contiguous |= (input.size(0) > 1 && input.stride(0) == 0);
+  if (need_contiguous) {
```

Comment:
In the documentation, it states:

> Out-of-place complex-to-real FFT will overwrite input buffer if custom strides are set by the user. 

Where is the test for this case?

Reply:
Added a test for this.